cli/src/persistent_key_store.rs

```diff
@@ -0,0 +1,135 @@
+        // hold the lock very briefly and should not panic during that period.
+        let values = self.values.read().unwrap();
+        if let Some(value) = values.get(k) {
+            serde_json::from_slice(value).ok()
+        } else {
+            None
+        }
+    }
+
+    /// Delete a value stored for ID `k`.
+    ///
+    /// Returns an error if storing fails.
+    fn delete<V: MlsEntity>(&self, k: &[u8]) -> Result<(), Self::Error> {
+        // We just delete both ...
+        let mut values = self.values.write().unwrap();
+        values.remove(k);
+        Ok(())
+    }
+}
+
+impl PersistentKeyStore {
+    fn get_file_path(user_name: &String) -> PathBuf {
+        return file_helpers::get_file_path(&("openmls_cli_".to_owned() + user_name + "_ks.json"));
+    }
+
+    fn save_to_file(&self, output_file: &File) -> Result<(), String> {
+        let writer = BufWriter::new(output_file);
+
+        let mut ser_ks = SerializableKeyStore::default();
+        for (key, value) in &*self.values.read().unwrap() {
+            ser_ks
+                .values
+                .insert(base64::encode(key), base64::encode(value));
+        }
+
+        match serde_json::to_writer_pretty(writer, &ser_ks) {
+            Ok(()) => Ok(()),
+            Err(e) => Err(e.to_string()),
+        }
+    }
+
+    pub fn save(&self, user_name: String) -> Result<(), String> {
+        let ks_output_path = PersistentKeyStore::get_file_path(&user_name);
+
+        match File::create(ks_output_path) {
+            Ok(output_file) => self.save_to_file(&output_file),
+            Err(e) => Err(e.to_string()),
+        }
+    }
+
+    fn load_from_file(&mut self, input_file: &File) -> Result<(), String> {
+        // Prepare file reader.
+        let reader = BufReader::new(input_file);
+
+        // Read the JSON contents of the file as an instance of `SerializableKeyStore`.
+        match serde_json::from_reader::<BufReader<&File>, SerializableKeyStore>(reader) {
+            Ok(ser_ks) => {
+                let mut ks_map = self.values.write().unwrap();
+                for (key, value) in ser_ks.values {
+                    ks_map.insert(base64::decode(key).unwrap(), base64::decode(value).unwrap());
+                }
+                Ok(())
+            }
+            Err(e) => Err(e.to_string()),
+        }
+    }
+
+    pub fn load(&mut self, user_name: String) -> Result<(), String> {
+        let ks_input_path = PersistentKeyStore::get_file_path(&user_name);
+
+        match File::open(ks_input_path) {
+            Ok(input_file) => self.load_from_file(&input_file),
+            Err(e) => Err(e.to_string()),
+        }
+    }
+}
+
+/// Errors thrown by the key store.
+#[derive(thiserror::Error, Debug, Copy, Clone, PartialEq, Eq)]
+pub enum PersistentKeyStoreError {
+    #[error("The key store does not allow storing serialized values.")]
+    UnsupportedValueTypeBytes,
+    #[error("Updating is not supported by this key store.")]
+    UnsupportedMethod,
```

Comment:
I'm getting clippy warnings that `UnsupportedValueTypeBytes` and `UnsupportedMethod` are never used, so they can probably go.